vm/src/obj/objstr.rs

```diff
@@ -1072,6 +1090,210 @@ fn call_object_format(vm: &VirtualMachine, argument: PyObjectRef, format_spec: &
+                            Err(vm.new_overflow_error("%c arg not in range(0x110000)".to_string()))
+                        }
+                    }
+                } else if objtype::isinstance(&obj, &vm.ctx.str_type()) {
+                    let s: String = get_value(&obj);
+                    let num_chars = s.chars().count();
+                    if num_chars != 1 {
+                        Err(vm.new_type_error("%c requires int or char".to_string()))
+                    } else {
+                        println!("Hurray!");
+                        Ok(s.chars().next().unwrap().to_string())
+                    }
+                } else {
+                    // TODO re-arrange this block so this error is only created once
+                    Err(vm.new_type_error("%c requires int or char".to_string()))
+                }
+            }?;
+            format_spec.precision = Some(CFormatQuantity::Amount(1));
+            Ok(format_spec.format_string(char_string))
+        }
+        _ => Err(vm.new_not_implemented_error(format!(
+            "Not yet implemented for %{}",
+            format_spec.format_char
+        ))),
+    }
+}
+
+fn try_update_quantity_from_tuple(
+    vm: &VirtualMachine,
+    elements: &mut Iterator<Item = PyObjectRef>,
+    q: &mut Option<CFormatQuantity>,
+    mut tuple_index: usize,
+) -> PyResult<usize> {
+    match q {
+        Some(CFormatQuantity::FromValuesTuple) => {
+            match elements.next() {
+                Some(width_obj) => {
+                    tuple_index += 1;
+                    if !objtype::isinstance(&width_obj, &vm.ctx.int_type()) {
+                        Err(vm.new_type_error("* wants int".to_string()))
+                    } else {
+                        // TODO: handle errors when truncating BigInt to usize
+                        *q = Some(CFormatQuantity::Amount(
+                            objint::get_value(&width_obj).to_usize().unwrap(),
+                        ));
+                        Ok(tuple_index)
+                    }
+                }
+                None => {
+                    Err(vm.new_type_error("not enough arguments for format string".to_string()))
+                }
+            }
+        }
+        _ => Ok(tuple_index),
+    }
+}
+
+fn do_cformat(
+    vm: &VirtualMachine,
+    mut format_string: CFormatString,
+    values_obj: PyObjectRef,
+) -> PyResult {
+    let mut final_string = String::new();
+    let num_specifiers = format_string
+        .format_parts
+        .iter()
+        .filter(|(_, part)| CFormatPart::is_specifier(part))
+        .count();
+    let mapping_required = format_string
+        .format_parts
+        .iter()
+        .any(|(_, part)| CFormatPart::has_key(part))
+        && format_string
+            .format_parts
+            .iter()
+            .filter(|(_, part)| CFormatPart::is_specifier(part))
+            .all(|(_, part)| CFormatPart::has_key(part));
+
+    let values_obj = if mapping_required {
```

Comment:
If `mapping_required` is used a lot, this might be a hint that there are actually two functions required. As a rule of thumb, if you have a function with a boolean as argument, just create two functions for the true and false case.

Reply:
Let's merge the current work in.
I will re-work the entire parsing-formatting later.
I have in mind to make CFormatString into an iterator and try to zip it with the tuple/dict like you suggested.